### PR TITLE
Updated PFCWD to use single ACL table for PFCWD and MUX

### PR DESCRIPTION
**Pull request by vmittal-msft**
**What I did**
Allow PFCWD and MUX features to use single ACL table on ingress side.
**Why I did it**
It is done to save space and to allow new features to make use of ACL table.
**How I verified it**
Verified by running tests on TD3 platform.
**Details if related**

### CODE REVIEW

File: orchagent/pfcactionhandler.cpp [hunk 1]
```diff
@@ -239,12 +240,39 @@ PfcWdAclHandler::PfcWdAclHandler(sai_object_id_t port, sai_object_id_t queue,
...
+            SWSS_LOG_NOTICE("Found the rule hence updating an existing one...%s:%s" PRIx64, m_strIngressTable.c_str(), m_strRule.c_str());
+            vector<sai_object_id_t> in_ports;
+            in_ports = ruleIter->second->getInPorts();
+            in_ports.push_back(port);
+
+            for (const auto& portIter: in_ports)
+            {
+                Port port;
+                gPortsOrch->getPort(portIter, port);
+                SWSS_LOG_NOTICE("Port number added to the list ................. %s" PRIx64, port.m_alias.c_str());
```
Comment (lguohan) [inline orchagent/pfcactionhandler.cpp:271]: syntax error.

File: orchagent/aclorch.cpp [hunk 2]
```diff
@@ -993,11 +993,26 @@ bool AclRuleL3::validate()
...
+    {
+        attr.id = *(sai_acl_entry_attr_t *)data;
+        attr.value = m_matches[*(sai_acl_entry_attr_t *)data];
+        attr.value.aclfield.enable = true;
+        
+        status = sai_acl_api->set_acl_entry_attribute(m_ruleOid, &attr);
+        if (status != SAI_STATUS_SUCCESS)
+        {
+            SWSS_LOG_ERROR("Failed to update ACL rule %s, rv:%d",
+                    m_id.c_str(), status);
```
Comment (prsunny) [inline orchagent/aclorch.cpp:1012]: This could be in the same line as above

File: orchagent/pfcactionhandler.cpp [hunk 3]
```diff
@@ -316,13 +354,42 @@ void PfcWdAclHandler::createPfcAclRule(shared_ptr<AclRulePfcwd> rule, uint8_t qu
...
 
+void PfcWdAclHandler::updatePfcAclRule(shared_ptr<AclRule> rule, uint8_t queueId, string strTable, vector<sai_object_id_t> in_ports)
+{
+    SWSS_LOG_ENTER();
+    string attr_name, attr_value;
+
+    for (const auto& portIter: in_ports)
+    {
+        Port p;
+        gPortsOrch->getPort(portIter, p);
```
Comment (prsunny) [inline orchagent/pfcactionhandler.cpp:382]: Suggest check for return value of this function call

File: orchagent/pfcactionhandler.cpp [hunk 4]
```diff
@@ -226,25 +226,42 @@ PfcWdAclHandler::PfcWdAclHandler(sai_object_id_t port, sai_object_id_t queue,
...
+    sai_object_id_t  table_oid;
 
     // There is one handler instance per queue ID
     string queuestr = to_string(queueId);
-    m_strIngressTable = "IngressTable_PfcWdAclHandler_" + queuestr;
+    m_strIngressTable = "IngressTable_PfcWdAclHandler";
     m_strEgressTable = "EgressTable_PfcWdAclHandler_" + queuestr;
     m_strRule = "Rule_PfcWdAclHandler_" + queuestr;
-
+
```
Comment (prsunny) [inline orchagent/pfcactionhandler.cpp:236]: tabspace?

**Comment by prsunny** [comment]
@vmittal-msft , as we discussed, can we have the in-port bitmap handling APIs generic, so that it can be used by other modules? Also you probably would want to revisit the table creation as it could be used before the PFC WD may start adding rules.

File: tests/test_pfcwd.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+def getQueueOid(dvs, queue):
+   cnt_r = redis.Redis(unix_socket_path=dvs.redis_sock, db=swsscommon.COUNTERS_DB,
+                       encoding="utf-8", decode_responses=True)
+   return  cnt_r.hget("COUNTERS_QUEUE_NAME_MAP", queue)
+
+def setPfcStormQueueOid(dvs, queue_oid):
+   cnt_r = redis.Redis(unix_socket_path=dvs.redis_sock, db=swsscommon.COUNTERS_DB,
+                       encoding="utf-8", decode_responses=True)
+   cnt_r.hset("COUNTERS:"+queue_oid, "DEBUG_STORM","enabled")
+   import pdb;pdb.set_trace()
```
Comment (daall) [inline tests/test_pfcwd.py:63]: `counters_db`

Comment (daall) [inline tests/test_pfcwd.py:63]: extra debug line

File: tests/test_pfcwd.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    atbl = swsscommon.Table(self.adb, "ASIC_STATE:SAI_OBJECT_TYPE_PORT")
+    port_groups = []
+    for p in [dvs.asicdb.portnamemap[portname] for portname in bind_ports]:
+        (status, fvs) = atbl.get(p)
+        for fv in fvs:
+            if fv[0] == "SAI_PORT_ATTR_INGRESS_ACL":
+                assert fv[1] in acl_table_groups
+                port_groups.append(fv[1])
+
+    assert len(port_groups) == len(bind_ports)
```
Comment (daall) [inline tests/test_pfcwd.py:132]: Take a look at `dvs_acl.py`, I think these methods are already implemented for you.

File: tests/test_pfcwd.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    assert set(port_groups) == set(acl_table_groups)
+
+def Verify_AclTableCreation(self, dvs):
```
Comment (daall) [inline tests/test_pfcwd.py:135]: I would take a look at https://github.com/Azure/sonic-swss/blob/0b0d24cfec42c0921687eb14cdb16f37adbfe34f/tests/test_acl.py#L54

File: tests/test_pfcwd.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    acl_group_ids = atbl.getKeys()
+    assert len(acl_group_ids) == 2
+
+    # check acl table group member
+    verify_acl_group_member(acl_group_ids, acl_table_id)
+
+    # check port binding
+    verify_acl_port_binding(dvs, "Ethernet0")
+
+def Verify_AclRuleInPorts(self, dvs):
```
Comment (daall) [inline tests/test_pfcwd.py:154]: I would take a look at https://github.com/Azure/sonic-swss/blob/0b0d24cfec42c0921687eb14cdb16f37adbfe34f/tests/test_acl.py#L101

File: tests/test_pfcwd.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    @pytest.fixture(scope="module", autouse=True)
+    def enable_queue_counters(self, dvs):
+        dvs.runcmd("counterpoll queue enable")
+
+    def test_ingress_pfcwd_handler(self, dvs):
+        self.setup_db(dvs)
+        config_db = dvs.get_config_db()
+        
+        #Enable PFC at global level
+        enable_device_pfcwd(config_db)
```
Comment (daall) [inline tests/test_pfcwd.py:189]: This can probably be made a fixture since it will be used throughout all of the tests.

**Review by daall — CHANGES_REQUESTED** [review]
as comments

File: orchagent/pfcactionhandler.cpp [hunk 10]
```diff
@@ -316,6 +354,22 @@ void PfcWdAclHandler::createPfcAclRule(shared_ptr<AclRulePfcwd> rule, uint8_t qu
     attr_value = to_string(queueId);
     rule->validateAddMatch(attr_name, attr_value);
 
+    // Add MATCH_IN_PORTS as match creteria for ingress table
+    if(strTable == INGRESS_TABLE_DROP)
```
Comment (prsunny) [inline orchagent/pfcactionhandler.cpp:358]: space after `if`

File: orchagent/pfcactionhandler.cpp [hunk 11]
```diff
@@ -225,35 +225,51 @@ PfcWdAclHandler::PfcWdAclHandler(sai_object_id_t port, sai_object_id_t queue,
...
         // First time of handling PFC for this queue, create ACL table, and bind
         createPfcAclTable(port, m_strIngressTable, true);
         shared_ptr<AclRulePfcwd> newRule = make_shared<AclRulePfcwd>(gAclOrch, m_strRule, m_strIngressTable, table_type);
-        createPfcAclRule(newRule, queueId, m_strIngressTable);
+        createPfcAclRule(newRule, queueId, m_strIngressTable, port);
     }
     else
     {
-        // Otherwise just bind ACL table with the port
-        found->second.bind(port);
```
Comment (daall) [inline orchagent/pfcactionhandler.cpp:247]: I think we still need to bind the ports to the ACL table before we update the rule so that the ACL group + group members are created for ports that don't exist in any other ACL tables.

File: orchagent/pfcactionhandler.cpp [hunk 12]
```diff
@@ -225,35 +225,51 @@ PfcWdAclHandler::PfcWdAclHandler(sai_object_id_t port, sai_object_id_t queue,
...
+        table_oid = gAclOrch->getTableById(m_strIngressTable);
+        shared_ptr<AclRulePfcwd> newRule = make_shared<AclRulePfcwd>(gAclOrch, m_strRule, m_strIngressTable, table_type);
```
[the remark this answers is not in the document]
Comment (vmittal-msft) [reply orchagent/pfcactionhandler.cpp:248]: Make sense. Done.

File: orchagent/aclorch.cpp [hunk 13]
```diff
@@ -2919,6 +2939,60 @@ bool AclOrch::removeAclRule(string table_id, string rule_id)
...
+            {
+                Port p;
+                gPortsOrch->getPort(port_iter, p);
+                attr_value += p.m_alias;
+                attr_value += ',';
+            }
+            attr_value.pop_back();
+
+            rule->validateAddMatch(MATCH_IN_PORTS, attr_value);  
+            m_AclTables[table_oid].rules[rule->getId()]->update(SUBJECT_TYPE_PORT_CHANGE, &aclMatchLookup[attr_name]);
```
Comment (neethajohn) [inline orchagent/aclorch.cpp:2985]: I do not understand why we are using SUBJECT_TYPE_PORT_CHANGE here. What is the change event here for the port? all we need to do here is call a function to set acl rule attributes.

File: orchagent/pfcactionhandler.cpp [hunk 14]
```diff
@@ -266,10 +282,23 @@ PfcWdAclHandler::~PfcWdAclHandler(void)
...
+    {
+        gAclOrch->removeAclRule(m_strIngressTable, m_strRule);
+    }
+    else 
+    {
+        gAclOrch->updateAclRule(rule_iter->second, m_strIngressTable, MATCH_IN_PORTS, &port, RULE_OPER_DELETE);
+    } 
+
+    auto found = m_aclTables.find(m_strEgressTable);
     found->second.unbind(getPort());
```
Comment (neethajohn) [inline orchagent/pfcactionhandler.cpp:302]: if Danny's comment on L247 holds good, then I think we need some extra handling for ingress table. what if the same port is bound to another rule for a diff queue?

**Review by neethajohn — COMMENTED** [review]
Haven't reviewed the vs test. will look at it when it is complete

File: orchagent/aclorch.cpp (hunk 13 above)
Comment (vmittal-msft) [reply orchagent/aclorch.cpp:2985]: ACL rule update is new function added as part of this PR. Subject type is a param used during rule update, I kept it to indicate different kind of rule updates we may be doing. Idea of passing SUBJECT_TYPE_PORT_CHANGE to indicate port mask is changing for the rule. Please let me know if i am incorrectly using subject type for this purpose.

File: orchagent/pfcactionhandler.cpp (hunk 14 above)
Comment (vmittal-msft) [reply orchagent/pfcactionhandler.cpp:302]: Sure. Will test out the L247 changes and make necessary changes here. Will probably unbind the port once there are no rules left in ingress table else we may nee do do walk of the table every time there is rule update.

File: orchagent/pfcactionhandler.cpp [hunk 15]
```diff
@@ -266,11 +283,30 @@ PfcWdAclHandler::~PfcWdAclHandler(void)
...
+    vector<sai_object_id_t> port_set = rule_iter->second->getInPorts();
+    sai_object_id_t port = getPort();
+
+    if ((port_set.size() == 1) && (port_set[0] == port))
+    {
+        gAclOrch->removeAclRule(m_strIngressTable, m_strRule);
+        if (table_size == 1) 
+        {
+            auto found = m_aclTables.find(m_strIngressTable);
+            found->second.unbind(port);
```
Comment (neethajohn) [inline orchagent/pfcactionhandler.cpp:300]: I am not sure if this is the right approach to unbind. we will need to unbind a port if its not used by any  rules in the table. @daall , @prsunny , can you confirm if we really need to bind the port to the table since now we are including it in the rule.

File: tests/test_pfcwd.py [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+            # Enable PFC Storm on queue 3
+            self.set_pfc_storm_queue_oid(dvs, queue_oid)
+            
+            # Verify ASIC_DB for table creation
+            #verify_acl_table_creation(self, dvs)
+
+            #Verify ASIC_DB for rule creation
+            #verify_acl_rule_in_ports(dvs)
+
+       finally:
```
Comment (neethajohn) [inline tests/test_pfcwd.py:289]: please make sure to clean up all the objects that you created for this particular test

File: orchagent/pfcactionhandler.cpp (hunk 15 above)
Comment (prsunny) [reply orchagent/pfcactionhandler.cpp:300]: Agree with @neethajohn, that this may cause issues to existing rules in that table.

File: orchagent/aclorch.cpp [hunk 17]
```diff
@@ -2919,6 +2949,63 @@ bool AclOrch::removeAclRule(string table_id, string rule_id)
...
+                attr_value += ',';
+            }
+            attr_value.pop_back();
+
+            rule->validateAddMatch(MATCH_IN_PORTS, attr_value);  
+            m_AclTables[table_oid].rules[rule->getId()]->update_in_ports(&aclMatchLookup[attr_name]);
+        }
+        break;
+
+        default:
```
Comment (daall) [inline orchagent/aclorch.cpp:3002]: It might be helpful to include a DEBUG log here in case someone tries to update a field that is not IN_PORTS

File: orchagent/aclorch.cpp [hunk 18]
```diff
@@ -1519,6 +1545,10 @@ bool AclTable::create()
         attr.id = SAI_ACL_TABLE_ATTR_FIELD_IP_PROTOCOL;
         attr.value.booldata = true;
         table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_IN_PORTS;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
```
Comment (daall) [inline orchagent/aclorch.cpp:1551]: How is this related to your PFCWD changes?

File: orchagent/pfcactionhandler.cpp [hunk 19]
```diff
@@ -295,14 +323,23 @@ void PfcWdAclHandler::createPfcAclTable(sai_object_id_t port, string strTable, b
...
     aclTable.link(port);
     aclTable.id = strTable;
-    aclTable.stage = ingress ? ACL_STAGE_INGRESS : ACL_STAGE_EGRESS;
+
+    if (ingress) 
+    {
+        aclTable.type = ACL_TABLE_DROP;
+        aclTable.stage = ACL_STAGE_INGRESS;
+    } else 
+    {
```
Comment (daall) [inline orchagent/pfcactionhandler.cpp:334]:
```suggestion
    } 
    else 
    {
```

File: tests/test_pfcwd.py [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+
+            expected_sai_qualifiers = {
+                "SAI_ACL_ENTRY_ATTR_FIELD_TC" : dvs_acl.get_simple_qualifier_comparator("4&mask:0xff"),
+                "SAI_ACL_ENTRY_ATTR_FIELD_IN_PORTS": dvs_acl.get_port_list_comparator(["Ethernet8","Ethernet24"]),
+            }
+
+            dvs_acl.update_acl_rule(PFCWD_TABLE_NAME, PFCWD_RULE_NAME_2, config_qualifiers, action="DROP")
+            time.sleep(5)
+            rule_oid = self.find_acl_rule_oid(dvs)
+            dvs_acl.verify_acl_rule(expected_sai_qualifiers, action="DROP",acl_rule_id=rule_oid)
```
Comment (daall) [inline tests/test_pfcwd.py:89]: IIRC if you leave off the OID it will automatically run a similar method to what you wrote in `find_acl_rule_oid`, so I think you can remove that method and just run this:

```suggestion
            dvs_acl.verify_acl_rule(expected_sai_qualifiers, action="DROP")
```

File: tests/test_pfcwd.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+
+            expected_sai_qualifiers = {
+                "SAI_ACL_ENTRY_ATTR_FIELD_TC" : dvs_acl.get_simple_qualifier_comparator("3&mask:0xff"),
+                "SAI_ACL_ENTRY_ATTR_FIELD_IN_PORTS": dvs_acl.get_port_list_comparator(["Ethernet0","Ethernet16"])
+            }
+
+            dvs_acl.update_acl_rule(PFCWD_TABLE_NAME, PFCWD_RULE_NAME_1, config_qualifiers, action="DROP")
+            time.sleep(5)
+            rule_oid = self.find_acl_rule_oid(dvs)
+            dvs_acl.verify_acl_rule(expected_sai_qualifiers, action="DROP",acl_rule_id=rule_oid)
```
Comment (daall) [inline tests/test_pfcwd.py:58]: same suggestion as below

File: tests/test_pfcwd.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+                "IN_PORTS": "Ethernet0,Ethernet16"
+            }
+
+            expected_sai_qualifiers = {
+                "SAI_ACL_ENTRY_ATTR_FIELD_TC" : dvs_acl.get_simple_qualifier_comparator("3&mask:0xff"),
+                "SAI_ACL_ENTRY_ATTR_FIELD_IN_PORTS": dvs_acl.get_port_list_comparator(["Ethernet0","Ethernet16"])
+            }
+
+            dvs_acl.update_acl_rule(PFCWD_TABLE_NAME, PFCWD_RULE_NAME_1, config_qualifiers, action="DROP")
+            time.sleep(5)
```
Comment (daall) [inline tests/test_pfcwd.py:56]: This sleep shouldn't be necessary as `verify_acl_rule` will continue to poll for the fields you specified.

File: tests/test_pfcwd.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+
+            expected_sai_qualifiers = {
+                "SAI_ACL_ENTRY_ATTR_FIELD_TC" : dvs_acl.get_simple_qualifier_comparator("3&mask:0xff"),
+                "SAI_ACL_ENTRY_ATTR_FIELD_IN_PORTS": dvs_acl.get_port_list_comparator(["Ethernet0"])
+            }
+        
+            dvs_acl.create_acl_rule(PFCWD_TABLE_NAME, PFCWD_RULE_NAME_1, config_qualifiers, action="DROP")
+            time.sleep(5)
+            rule_oid = self.find_acl_rule_oid(dvs)
+            dvs_acl.verify_acl_rule(expected_sai_qualifiers, action="DROP", acl_rule_id=rule_oid)
```
Comment (daall) [inline tests/test_pfcwd.py:43]: same feedback as below w.r.t. sleep + verify_acl_rule

File: tests/test_pfcwd.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+       keys = atbl.getKeys()
+       acl_entry = [k for k in keys if k not in dvs.asicdb.default_acl_entries]
+
+       return acl_entry[0]
+
+    def test_PfcWdAclCreationDeletion(self, dvs, dvs_acl, testlog):
+        try:
+            dvs_acl.create_acl_table(PFCWD_TABLE_NAME, PFCWD_TABLE_TYPE, ["Ethernet0","Ethernet8", "Ethernet16", "Ethernet24"], stage="ingress")
+            dvs.runcmd("counterpoll queue enable")
+            time.sleep(10)
```
Comment (daall) [inline tests/test_pfcwd.py:28]: This is OK for this PR, but it would be better to poll for some event (like the QUEUE_NAME_MAP appearing in the DB) that indicates this command was successful rather than adding a static 10s sleep.

Comment (daall) [reply tests/test_pfcwd.py:28]: I also don't know if this step is still necessary for this PR or not given that we're going through config DB directly instead of triggering the PFCWD.

File: tests/test_pfcwd.py (hunk 22 above)
Comment (vmittal-msft) [reply tests/test_pfcwd.py:56]: Without sleep, it seems to be failing.

File: tests/test_pfcwd.py (hunk 23 above)
Comment (vmittal-msft) [reply tests/test_pfcwd.py:43]: Fixed verify_acl_rule but without sleep it seems to be failing.

File: tests/test_pfcwd.py (hunk 22 above)
Comment (daall) [reply tests/test_pfcwd.py:45]: OK I don't want to block this PR any further so let's revisit it later. :)

**Review by daall — APPROVED** [review]

File: orchagent/aclorch.h [hunk 25]
```diff
@@ -196,6 +201,7 @@ class AclRule
     virtual bool create();
     virtual bool remove();
     virtual void update(SubjectType, void *) = 0;
+    virtual void update_in_ports(void *);
```
Comment (prsunny) [inline orchagent/aclorch.h:204]: please use same naming convention for the function as the rest of class

File: orchagent/aclorch.cpp [hunk 26]
```diff
@@ -620,6 +621,26 @@ bool AclRule::remove()
     return res;
 }
 
+void AclRule::update_in_ports(void *data)
```
Comment (prsunny) [inline orchagent/aclorch.cpp:624]: Here its not required to be a void * param as the API is specific for inports

File: orchagent/aclorch.cpp [hunk 27]
```diff
@@ -2919,6 +2945,64 @@ bool AclOrch::removeAclRule(string table_id, string rule_id)
...
+            attr_value.pop_back();
+
+            rule->validateAddMatch(MATCH_IN_PORTS, attr_value);  
+            m_AclTables[table_oid].rules[rule->getId()]->update_in_ports(&aclMatchLookup[attr_name]);
+        }
+        break;
+
+        default:
+            SWSS_LOG_ERROR("Acl rule update not supported for attr name %s", attr_name.c_str());
+            break;
```
Comment (prsunny) [inline orchagent/aclorch.cpp:3000]: please align break statement similar to other case

**Comment by vmittal-msft** [comment]
retest vs please

**Review by prsunny — APPROVED** [review]
lgtm.
